src/librustc_mir/interpret/intrinsics.rs

```diff
@@ -435,6 +440,41 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         Ok(true)
     }
 
+    fn guaranteed_eq(&mut self, a: Scalar<M::PointerTag>, b: Scalar<M::PointerTag>) -> bool {
+        // `Some` in `cmp` means the comparison is known, guaranteed in both directions
+        match (a, b) {
+            // Comparisons between integers are always known.
+            (Scalar::Raw { .. }, Scalar::Raw { .. }) => a == b,
+            // Equality with integers can never be known for sure.
+            (Scalar::Raw { .. }, Scalar::Ptr(_)) | (Scalar::Ptr(_), Scalar::Raw { .. }) => false,
+            // FIXME: return `true` for at least some comparisons where we can reliably
+            // determine the result of runtime equality tests at compile-time.
+            // Examples include comparison of addresses in static items, for these we can
+            // give reliable results.
```

Comment:
The easiest example is two identical pointers. They are certainly guaranteed equal. And in fact I think that's as good as we can be -- `static` items help for guaranteed inequality but not for guaranteed equality.

Reply:
two abstract pointers to the same function are not guaranteed to be equal at runtime, while at compile-time they are, so that's not true. I thought there were pointers to stack where that could also hold (llvm optimizations may duplicate local variables theoretically)

Reply:
Two CTFE-created pointers to the same function should be equal at runtime -- we are creating the exact same LLVM IR for them.

But okay, we can punt on this for now. Just the comment is misleading IMO -- talking about statics makes sense for the "ne" case but not for the "eq" case. Please talk explicitly about things that are equal at CTFE but could have different addresses at runtime, such as functions or vtables.

>  I thought there were pointers to stack where that could also hold (llvm optimizations may duplicate local variables theoretically)

I don't think LLVM optimizations can duplicate local variables you have a pointer to. And CTFE cannot even point to runtime stack variables.